compiler/rustc_typeck/src/check/method/prelude2021.rs

```diff
@@ -206,6 +210,32 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             return;
         }
 
+        // For from_iter, check if the type actualy implements FromIterator.
+        // If we know it does not, we don't need to warn.
+        if method_name.name == sym::from_iter {
+            if let Some(trait_def_id) = self.tcx.lang_items().from_iterator_trait() {
+                if self.tcx.infer_ctxt().enter(|infcx| {
+                    let obligation = Obligation::new(
+                        ObligationCause::dummy(),
+                        ty::ParamEnv::empty(),
+                        ty::PredicateKind::Trait(
+                            ty::TraitPredicate {
+                                trait_ref: ty::TraitRef {
+                                    def_id: trait_def_id,
+                                    substs: self.tcx.mk_substs_trait(self_ty, &[]),
+                                },
+                            },
+                            Constness::NotConst,
+                        )
+                        .to_predicate(self.tcx),
+                    );
+                    !infcx.predicate_may_hold(&obligation)
```

Comment:
this block can be simplified with `tcx.type_implements_trait(...)`
https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/query/trait.QueryEngine.html#tymethod.type_implements_trait